### PR TITLE
fix state dict to save/load learning rate scheduler

### PR DESCRIPTION
**Pull request by zhwesky2010**
### PR types

Function optimization
### PR changes

APIs
### Describe

1.8新增了多种与Pytorch对齐的Decay策略，属于动态图下epoch刷新（已合入4种 ReduceLROnPlateau、StepDecay、MultiStepDecay、LambdaDecay，未来仍可能继续增加，需要有较好的基类以供复用），与之前策略的成员变量有区别，因此目前 ``save_dygraph/load_dygraph`` 不可用。

该PR拓展、优化了基类 ``class Optimizer`` 与基类 ``class LearningRateDecay`` 的``state_dict``、``set_dict``函数，并可以适用于全部的Decay策略，包含以下三种：

- 按step更新的学习率策略，例如 NoamDecay 、 PiecewiseDecay 、NaturalExpDecay等，默认保存：
 ``step_num``

- 按epoch更新的学习率策略，例如 StepDecay 、 MultiStepDecay 、 LambdaDecay等，默认保存：
``epoch_num, learning_rate``

- ReduceLROnPlateau策略，该策略为特殊策略，中间变量很多，需保存：
``cooldown_counter, best_loss, num_bad_epochs, epoch_num, learning_rate``

- 如继续新增竞品对齐策略，不走默认的保存的话，仅重写
```
def _state_keys(self):
    self.keys = ['step_num'] 
```
  即可。

### CODE REVIEW

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py [hunk 1]
```diff
@@ -66,6 +66,42 @@ def create_lr_var(self, lr):
...
+        """
+        set the keys in self.__dict__ that are needed to be saved.
+        """
+        self.keys = ['step_num']
+
+    def set_dict(self, state_dict):
+        """
+        Loads the schedulers state.
+        """
+        self.__dict__.update(state_dict)
```
Comment (phlrain) [inline python/paddle/fluid/dygraph/learning_rate_scheduler.py]: 这个是不是要根据 keys 进行update

Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py]: state_dict只会生成self.keys对应的dict。加载的这个就是self.keys对应的state_dict。这个需要一个个判断keys的内容吗

Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py]: Done

**Review by lanxianghit — APPROVED** [review]

**Review by swtkiwi — APPROVED** [review]
LGTM
